### PR TITLE
avoid redundant memory allocation for external initializers

### PR DESCRIPTION
**Pull request by frank-dong-ms**
### Description
avoid redundant memory allocation for external initializers, we will use mmap for external initializers later so no point to allocate memory in advance then release them later.

### Motivation and Context
In current implementation, we will:
1.	Allocate memory (with desired size of current initializer) for initializer first: [https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/framework/session_state_utils.cc#L131](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fonnxruntime%2Fblob%2Fmain%2Fonnxruntime%2Fcore%2Fframework%2Fsession_state_utils.cc%23L131&data=05%7C02%7Cfrdong%40microsoft.com%7C1e126797c95149aa217d08dcb781cc60%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638587015340041125%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=6fN57MUsergrCX%2BBS7jztWBRmc8nx19EVvn0lUJ2Gtk%3D&reserved=0)
2.	For external initializer, we will point initializer to mmaped object in memory and release previously allocated tensor: [https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/framework/session_state_utils.cc#L89](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fonnxruntime%2Fblob%2Fmain%2Fonnxruntime%2Fcore%2Fframework%2Fsession_state_utils.cc%23L89&data=05%7C02%7Cfrdong%40microsoft.com%7C1e126797c95149aa217d08dcb781cc60%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638587015340054491%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C0%7C%7C%7C&sdata=yBtXLc%2Bhpx3IT1%2FX0664foqQ5X5O%2Fy5XNhj4Oed%2BAt4%3D&reserved=0)

For large models, we are keep allocating and release memory for external initializers which seems unnecessary.

For phi silica model, with this change we can reduce transient memory usage from 4,566MB to 2,724MB. Since these redundant memory is released quickly when we mmap external initializers so this change has no much impact on peak memory usage.

### CODE REVIEW

File: onnxruntime/core/framework/session_state_utils.cc [hunk 1]
```diff
@@ -156,15 +162,21 @@ static common::Status DeserializeTensorProto(const Env& env, const std::basic_st
 
     // deserialize to CPU first for non-CPU allocator, then copy
     std::unique_ptr<Tensor> p_deserialize_tensor;
-    if (use_device_allocator_for_initializers) {
-      void* tensor_buffer = nullptr;
-      ORT_RETURN_IF_ERROR(AllocateBufferUsingDeviceAllocatorFromShapeAndType(tensor_shape, type, default_cpu_alloc, tensor_buffer));
-      p_deserialize_tensor = std::make_unique<Tensor>(type, tensor_shape, tensor_buffer, default_cpu_alloc);
+
+    // for external initializer we will use mmap so don't need to allocate memory in advance
+    if (!utils::HasExternalData(tensor_proto)) {
```
Comment (pranavsharma) [inline onnxruntime/core/framework/session_state_utils.cc:167]: Didn't realize that this was repeated here. This function ```DeserializeTensorProto``` is getting really hard to read. Potential for cleanup.

**Review by pranavsharma — DISMISSED** [review]

File: onnxruntime/core/framework/session_state_utils.cc (hunk 1 above)
Comment (frank-dong-ms) [reply onnxruntime/core/framework/session_state_utils.cc:167]: Yes, I have discussion with Yulong today about this part as well, I will re-organize this part.

**Review by guschmue — DISMISSED** [review]

File: onnxruntime/core/framework/session_state_utils.cc [hunk 2]
```diff
@@ -113,28 +113,14 @@ static common::Status DeserializeTensorProto(const Env& env, const std::basic_st
...
-      // (may expand it if there isn't a chunk that can be allotted to the memory request).
-      // If the provided allocator is non-arena based, the device specific Alloc() call will be used to allocate the necessary memory.
-      p_tensor = std::make_unique<Tensor>(type, tensor_shape, alloc);
-    }
-  }
 
-  if (p_tensor->Location().device.Type() == OrtDevice::CPU) {
-    // deserialize directly to CPU tensor
-    if (utils::HasExternalData(tensor_proto)) {
+  auto device_type = (alloc != nullptr) ? alloc->Info().device.Type() : m->GetAllocInfo().device.Type();
```
Comment (fs-eire) [inline onnxruntime/core/framework/session_state_utils.cc:117]: is it possible that `m` is `nullptr`?

Comment (frank-dong-ms) [reply onnxruntime/core/framework/session_state_utils.cc:117]: m can be nullptr but it is not possible both alloc and m are nullptr (see line 104-107 for assertion)

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**